util.js

```diff
@@ -251,6 +251,56 @@ var connectSmart = function() {
         return connectSmart3A.apply(null, arguments)
 }
 
+/**
+ * Helper class to wait for the system startup to be completed.
+ * Adding widgets before the desktop is ready to accept them can result in errors.
+ */
+var StartUpCompletionHelper = class AppIndicators_StartUpCompletionHelper {
+
+    constructor() {
+        this._startupComplete = false;
+        this._displayAvailable = false;
+        this._listeners = new Set();
+
+        // If the desktop is still starting up, we must wait until it is ready
+        if (Main.layoutManager._startingUp) {
+            let startupPreparedId = Main.layoutManager.connect('startup-complete', () => {
+                Main.layoutManager.disconnect(startupPreparedId);
+                this._startupComplete = true;
+                this._checkStartUpComplete();
+            });
+        } else {
+            this._startupComplete = true;
+        }
+
+        // Ensure that the default Gdk Screen is available
+        if (Gtk.IconTheme.get_default() == null) {
+            let waitForThemeId = Gdk.DisplayManager.get().connect('display-opened', () => {
+                Gdk.DisplayManager.get().disconnect(waitForThemeId);
+                this._displayAvailable = true;
+                this._checkStartUpComplete();
+            });
+        } else {
+            this._displayAvailable = true;
+        }
+    }
+
+    _checkStartUpComplete() {
+        if (this._startupComplete && this._displayAvailable) {
+            this._listeners.forEach(callback => callback());
+            this._listeners.clear();
+        }
+    }
+
+    whenStartUpComplete(callback) {
+        if (this._startupComplete && this._displayAvailable) {
+            callback();
+        } else {
```

Comment:
We can delete else and just use this._listeners.add(callback)

Reply:
Maybe I am misunderstanding your intention.
The check to invoke it immediately is required for all registrations that happen after the startup completed.
Always adding the callback to the list of listeners would result in an endless growing list that is never used again.